Libraries/LibWeb/CSS/Parser/PropertyParsing.cpp

```diff
@@ -1658,6 +1662,120 @@ RefPtr<CSSStyleValue const> Parser::parse_border_value(PropertyID property_id, T
 }
 
+RefPtr<CSSStyleValue const> Parser::parse_border_image_value(TokenStream<ComponentValue>& tokens)
+{
+    // <'border-image-source'> || <'border-image-slice'> [ / <'border-image-width'> | / <'border-image-width'>? / <'border-image-outset'> ]? || <'border-image-repeat'>
+    auto transaction = tokens.begin_transaction();
+
+    RefPtr<CSSStyleValue const> source_value;
+    RefPtr<CSSStyleValue const> slice_value;
+    RefPtr<CSSStyleValue const> width_value;
+    RefPtr<CSSStyleValue const> outset_value;
+    RefPtr<CSSStyleValue const> repeat_value;
+
+    auto make_border_image_shorthand = [&]() {
+        transaction.commit();
+        if (!source_value)
+            source_value = property_initial_value(PropertyID::BorderImageSource);
+        if (!slice_value)
+            slice_value = property_initial_value(PropertyID::BorderImageSlice);
+        if (!width_value)
+            width_value = property_initial_value(PropertyID::BorderImageWidth);
+        if (!outset_value)
+            outset_value = property_initial_value(PropertyID::BorderImageOutset);
+        if (!repeat_value)
+            repeat_value = property_initial_value(PropertyID::BorderImageRepeat);
+
+        return ShorthandStyleValue::create(PropertyID::BorderImage,
+            { PropertyID::BorderImageSource, PropertyID::BorderImageSlice, PropertyID::BorderImageWidth, PropertyID::BorderImageOutset, PropertyID::BorderImageRepeat },
+            { source_value.release_nonnull(), slice_value.release_nonnull(), width_value.release_nonnull(), outset_value.release_nonnull(), repeat_value.release_nonnull() });
+    };
+
+    auto parse_value_list = [&](PropertyID property_id, TokenStream<ComponentValue>& inner_tokens, Optional<char> delimiter = {}) -> RefPtr<CSSStyleValue const> {
+        auto inner_transaction = inner_tokens.begin_transaction();
+
+        auto remaining_values = property_maximum_value_count(property_id);
+        StyleValueVector values;
+        while (inner_tokens.has_next_token() && remaining_values > 0) {
+            if (delimiter.has_value() && inner_tokens.next_token().is_delim(*delimiter))
+                break;
+            auto value = parse_css_value_for_property(property_id, inner_tokens);
+            if (!value)
+                break;
+            values.append(*value);
+            --remaining_values;
+        }
+        if (values.is_empty())
+            return nullptr;
+
+        inner_transaction.commit();
+        if (values.size() == 1)
+            return values[0];
+
+        return StyleValueList::create(move(values), StyleValueList::Separator::Space);
+    };
+
+    auto parse_repeat_value = [&]() {
+        tokens.discard_whitespace();
+        if (auto value = parse_value_list(PropertyID::BorderImageRepeat, tokens)) {
+            repeat_value = move(value);
+            return true;
+        }
+        return false;
+    };
+
+    if (parse_repeat_value())
+        return make_border_image_shorthand();
```

Comment:
This doesn't look right. That's a `||` before `<'border-image-repeat'>`, so it can go before or between the other two parts.